core/src/identifier.rs

```diff
@@ -240,9 +240,11 @@ mod interner {
     /// The interner, which serves a double purpose: it pre-allocates space
     /// so that [Ident](super::Ident) labels are created faster
     /// and it makes it so that labels are stored only once, saving space.
-    pub(crate) struct Interner<'a>(RwLock<InnerInterner<'a>>);
+    // NOTE: We set the lifetime parameter of InnerInterner to 'static since
+    // it's arbitrary, and there's no reason to expose it to users of Interner
+    pub(crate) struct Interner(RwLock<InnerInterner<'static>>);
```

Comment:
@Radvendii , but it seems that it USED to be the fact that the lifetime of the interner and the inner interner were the same one, before this change. Or am I misunderstanding something? Does your first example of UB still happens with the original code?

Reply:
Yes, it still happens with the original code. 

Remember that in `Interner<'a>(RwLock<InnerInterner<'a>)` , there is nothing limiting `'a` to the liferime of the Interner. If rust needs it to be `'static` it will just insert `'static` in here anyways, so specifying `'static` doesn't allow any behaviour that wasn't allowed before.

Reply:
Ah, I see. But it's a private item already, as is `InnerInterner`, so you can only access the `InnerInterner` of an interner from within the `identifier::interner` module, is that right?

Reply:
in fact, I start to believe that the lifetime of `InnerInterner` is useless, because we only ever instantiate it to static. So maybe we should just bite the bullet and use `static` references, saying that they aren't static at all, but only the actual implementation of `InnerInterner` can touch it. And document as well that an `Interner` should never ever change the underlying inner interner after construction. However, the inner interner being private should make it relatively ok. Another solution is to inline the inner interner in the `Interner` to avoid this situation, but this means more locks, and more coupling. I don't know if this is worth the price: in any case, we should make it clear that this module has unsafe parts and that you shouldn't touch it unless you know what you're doing.

Reply:
Another possibility is to explore `Pin`, which believe allows to express those kind of constraints (for example you can't swap a `Pin<..>` with something else, just drop it, which might be sufficient to avoid your UB example ?)

Reply:
And finally, maybe we should use `NonNull` instead of references with lifetimes, which would avoid to lie about their lifetime (would still require unsafe, but at least won't make it possible to do conversion to `&'static` from safe code)